File: DSA PRACTICE/LECTURE 22/ValidPalindrome.java
// PROGRAM 05: Valid Palindrome (125 leetcode and Code studio) 
/*
STEP01: faltu character remove karne hai
STEP02: sab lowercase me kardo
STEP03: check kro palindrome hai ya nhi
*/
import java.util.* ;
import java.io.*; 
class Solution 
{
    public boolean isPalindrome(String s) 
    {
        // STEP01 AND STEP02
        s = s.toLowerCase().replaceAll("[^A-Za-z0-9]", "");

        //STEP03
        int i = 0;
        int j = s.length() - 1;
        while(i <= j) 
        {
            if (s.charAt(i) != s.charAt(j)) 
            {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

/*
Example 1:
Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.

Example 2:
Input: s = "race2 a car"
Output: false
Explanation: "race2acar" is not a palindrome.

Example 3:
Input: s = " "
Output: true
Explanation: s is an empty string "" after removing non-alphanumeric characters.
Since an empty string reads the same forward and backward, it is a palindrome.
*/

File: DSA PRACTICE/LECTURE 22/ValidPalindrome.cpp
// PROGRAM 05: Valid Palindrome (125 leetcode and Code studio) 

class Solution {
   
private:
    bool valid(char ch)
    {
        if((ch>='a'&&ch<='z')||(ch>='A'&&ch<='Z')||(ch>='0'&&ch<='9'))
        {
            return 1;
        }
        return 0;
    }
    
    char toLowerCase(char ch) 
    {
        if(ch >='a' && ch <='z')
        {
            return ch;
        }
        else
        {
            char temp = ch - 'A' + 'a';
            return temp;
        }
    }
    
    bool checkPalindrome(string temp) 
    {
        int s = 0;
        int e = temp.length()-1;

        while(s<=e) 
        {
            if(temp[s] != temp[e])
            {
                return 0;       
            }
            else
            {
                s++;
                e--;
            }
        }
        return 1;
    }
    
public:
    bool isPalindrome(string s) 
    {
        //faltu character remove karne hai
        string temp="";
        
        for(int j=0;j<s.length();j++)
        {
            if(valid(s[j]))
            {
                temp.push_back(s[j]);
            }
        }
        
        
        //sab lowercase me kardo
        for(int j=0;j<s.length();j++)
        {
            temp[j]=toLowerCase(temp[j]);
        }
        
        //check kro palindrome hai ya nhi
        return checkPalindrome(temp); 
        
    }
};
/*
Example 1:
Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.

Example 2:
Input: s = "race2 a car"
Output: false
Explanation: "race2acar" is not a palindrome.

Example 3:
Input: s = " "
Output: true
Explanation: s is an empty string "" after removing non-alphanumeric characters.
Since an empty string reads the same forward and backward, it is a palindrome.
*/